gh-60712: Include the "object" type in the lists of documented types (#103036)
python/cpython: Doc/reference/datamodel.rst

   pair: extension; module
   pair: C; language

Below is a list of the types that are built into Python. The fundamental type
is :class:`object`; all other types, whether built-in or not, are derived
from it. Extension modules (written in C, Java, or other languages,
depending on the implementation) can define additional types.  Future versions
of Python may add types to the type hierarchy (e.g., rational numbers,
efficiently stored arrays of integers, etc.), although such additions will

> often be provided via the standard library instead.

Reviewer: Isn't this a little out of date? rational numbers = `Fraction`, efficiently stored arrays of integers = ``array``? And weren't these (and all likely future additions) via the standard library rather than built in to the language?

Reviewer 2: Agreed.  I will change this.

   pair: extension; module
   pair: C; language

Below is a list of the types that are built into Python. The fundamental type
is :class:`object`; all other types, whether built-in or not, are derived
from it. Extension modules (written in C, Java, or other languages,
depending on the implementation) can define additional types.  Future versions

> of Python may add types to the type hierarchy (e.g., rational numbers,

Reviewer 3: This sentence feels like it was written decades ago and we may be better taking it out. Not necessarily in scope for this PR though.

Reviewer 4: I had the same thought 🙂  We do have fractions now, integer arrays…

   class).  The return value of :meth:`__new__` should be the new object instance
   (usually an instance of *cls*).

   Typical custom implementations create a new instance of the class by invoking the
   superclass's :meth:`__new__` method using ``super().__new__(cls[, ...])``

>    with appropriate arguments and then modify the newly-created instance as necessary

Reviewer 4: I don’t think this needs to be changed: adjective + adjective compounds use a hyphen, but adverb + adjective don’t need it.

   finalizer or (improperly) a destructor.  If a base class has a
   :meth:`__del__` method, the derived class's :meth:`__del__` method,
   if any, must explicitly call it to ensure proper deletion of the base

>    class part of the instance.The :class:`object` class itself does not have a

Reviewer: ```suggestion
   class part of the instance. The :class:`object` class itself does not have a
```

Missing space (typo)

.. method:: object.__str__(self)

   Called by :func:`str(object) <str>`, the default :meth:`__format__` implementation,
   and the built-in function :func:`print`, to compute the "informal" or nicely
   printable string representation of an object.  The return value must be a

>    :ref:`string <textseq>` object.

Reviewer: While we're at it, why not cross-reference the actual class (with the actual name and correct format) rather than just the general docs section.

```suggestion
   :class:`str` object.
```

   The return value must be a string object.

>    The default implementation by the ``object`` class should be given

Reviewer: ```suggestion
   The default implementation by the :class:`object` class should be given
```

Resolve the reference since we're not actually inside `object` proper right now, and to get consistent formatting and semantics.

>    an empty ``format_spec`` string. It delegates to :meth:`__str__`

Reviewer: ```suggestion
   an empty *format_spec* string. It delegates to :meth:`__str__`
```

Function parameters are in *itlatic*, by convention.

   ``(x<y or x==y)`` does not imply ``x<=y``. To automatically generate ordering
   operations from a single root operation, see :func:`functools.total_ordering`.

   By default, the :class:`object` class provides implementations consistent
   with :ref:`comparisons`: equality compares according to object identity,
   and order comparisons raise :exc:`TypeError`. Each default method may generate

>    these results directly, but may also return ``NotImplemented``.

Reviewer: ```suggestion
   these results directly, but may also return :data:`NotImplemented`.
```

IMO, the previous mentions (or at least the first instance) should be updated, but isn't strictly necessary in this PR.

   class).  The return value of :meth:`__new__` should be the new object instance
   (usually an instance of *cls*).

>    Typical custom implementations create a new instance of the class by invoking the

Reviewer 4: Disagree with this addition, it does not clarify or improve what is said.

The line is about typical implementations of a `__new__` method, meaning any code written in stdlib or by other developers. «custom» does not make sense here; we are not talking about «custom classes» (meaning non-builtin), but literally about writing the code for the method in your class.

Reviewer 5: This section is the language reference so we want to stay lean on implementation details.

I would leave this section as originally written.

Reviewer 6: I think I was pointing out this sentence is about implementations in other classes, not in _object_ itself. But maybe it doesn’t need to be explicit, especially if the next sentence about the base implementation stays (or if specification of the concrete _object_ type is moved to e.g. library/stdtypes.rst). Agree I picked a confusing word.

Reviewer 4: Oh, this whole section is about writing your `__new__` method in your python class, not `object` itself! (line 1599 is confusing)

Reviewer 4: ```suggestion
   Typical implementations create a new instance of the class by invoking the
```

   superclass's :meth:`__new__` method using ``super().__new__(cls[, ...])``

>    with appropriate arguments and then modify the newly created instance as necessary

Reviewer 5: Please leave as originally written. Thanks!

   Called after the instance has been created (by :meth:`__new__`), but before
   it is returned to the caller.  The arguments are those passed to the
   class constructor expression.  If a base class has an :meth:`__init__`

>    method, the derived class's :meth:`__init__` method, if any, may need to

Reviewer 5: Please use the original wording.

Reviewer 6: But despite _object_ having an `__init__` method, I think it’s safe to assume the instance is “properly initialized” without a derived `__init__` calling `object.__init__`

Reviewer 4: The text is about `SomeClass.__init__` _very probably_ needing to call `SuperClass.__init__` to work properly, it’s not about `object.__init__` (which does nothing).

   call it to ensure proper initialization of the base class part of the
   instance; for example: ``super().__init__([args...])``.

>    Any :meth:`__init__` method that is overridden is not implicitly called.

Reviewer 4: This feels redundant with the previous sentence + that’s not specific to `__init__`

Reviewer 5: Please remove the added implementation details. They are unnecessary in the language reference document.

   finalizer or (improperly) a destructor.  If a base class has a
   :meth:`__del__` method, the derived class's :meth:`__del__` method,
   if any, must explicitly call it to ensure proper deletion of the base

>    class part of the instance. The :class:`object` class itself does not have a

Reviewer 5: Please remove additional text.

> .. method:: object.__str__(self)

Reviewer 4: I thought we had a doc bug years ago where we came up with a convention to differentiate `object.__str__` (actual method that exists in the `object` class), `anyclass.__bases__` (method that exists on all class; explaining the data model) and `someclass.__str__` (method you can define in your class to implement a protocol)

I guess we didn’t…

   The return value must be a string object.

   The default implementation by the :class:`object` class should be given

>    an empty *format_spec* string. It delegates to :meth:`__str__`

Reviewer 6: Missing full stop (.) at end of paragraph.

   ``(x<y or x==y)`` does not imply ``x<=y``. To automatically generate ordering
   operations from a single root operation, see :func:`functools.total_ordering`.

   By default, the :class:`object` class provides implementations consistent

>    with :ref:`comparisons`: equality compares according to object identity,

Reviewer 6: Can we link direct to :ref:`` `expressions-value-comparisons` ``? The general text directly under the Comparisons heading is not relevant.

   and order comparisons raise :exc:`TypeError`. Each default method may generate

>    these results directly, but may also return :data:`NotImplemented`.

Reviewer 6: Since the full behaviour of both `__eq__` and `__ne__` is now specified in the paragraph above, perhaps reduce this to “The four default order comparison methods may raise TypeError directly, or may instead return NotImplemented.”

.. method:: object.__dir__(self)

   Called when :func:`dir` is called on the object. An iterable must be
   returned. The :func:`dir` function converts the returned iterable to a list

>    and sorts it. The :class:`object` class itself provides an implementation

Reviewer 5: I would recommend staying with the original wording as it is more concise.

.. method:: object.__dir__(self)

   Called when :func:`dir` is called on the object. An iterable must be

>    returned. The :func:`dir` converts the returned iterable to a list and

Reviewer 6: Missing noun: The :func:`` `dir` `` _function_ converts . . .

.. method:: object.__str__(self)

>    Called by :func:`str(object) <str>`, the default :meth:`__format__` implementation,

Reviewer 4: Does this role (without class reference) create a proper link to the \_\_format\_\_ method doc?

Author: When you click on   ```:meth:`__format__```, it directs you to ```object.__format__(self, format_spec)```

   and the built-in function :func:`print`, to compute the "informal" or nicely
   printable string representation of an object.  The return value must be a

>    :class:`str` object.

Reviewer 4: The link went nicely to the `str` section, a few paragraph above the target of this changed link (the str function definition).

If the goal of the change is to say «str» and not «string», to avoid imprecision or suggesting that there is more than one string type, I see the value! Then let’s do this:

```suggestion
   :ref:`str <textseq>` object.
```

descriptor must be in either the owner's class dictionary or in the class
dictionary for one of its parents).  In the examples below, "the attribute"
refers to the attribute whose name is the key of the property in the owner

> class' :attr:`__dict__`.  The :class:`object` class itself does not

Reviewer 6: Does the link to `__dict__` still work? It looks like you are reverting a change from revision bae5d81f5d1.

   immutable (if the object's hash value changes, it will be in the wrong hash
   bucket).

>    User-defined classes (and the :class:`object` class itself) have :meth:`__eq__`
>    and :meth:`__hash__` methods by default; with them, all objects compare

Reviewer 4: Is this saying:
```suggestion
   User-defined classes have :meth:`__eq__` and :meth:`__hash__` methods
   by default (inherited from the :class:`object` class); with them, all objects compare
```

   Called when the instance is "called" as a function; if this method is defined,
   ``x(arg1, arg2, ...)`` roughly translates to ``type(x).__call__(x, arg1, ...)``.
   The :class:`object` class itself does not provide this method.

.. _sequence-types:

Emulating container types
-------------------------

The following methods can be defined to implement container objects. None of them
are provided by the :class:`object` class itself. Containers usually are
:term:`sequences <sequence>` (such as :class:`lists <list>` or
:class:`tuples <tuple>`) or :term:`mappings <mapping>` (like

> :class:`dictionaries <dict>`),

Reviewer: ```suggestion
:class:`dicts <dict>`),
```

This is emitted as a literal, so it should be the literal name (or very close to it)

Reviewer 4: Friendly disagreement here!  IMO it’s fine to have the class formatting even if we’re using the full term _dictionaries_.
It’s a normal word to read, we get a link to more info, the formatting is not a problem.

Reviewer: Friendly disagreement with the friendly disagreement here :) Since this is displayed as a literal, that by definition implies that it is the _literal_ code name of the class referenced, which is of course not the case. However, its not the end of the world; we can agree to disagree here and leave it to @terryjreedy to decide either way?

Reviewer 4: I think your approach is to ask whether the markup is 100% philosophically correct.
Mine is to ask whether there is a problem (in function, like links, or confusion for readers), which I really don’t think there is with literal style on a non-code term here.

Reviewer: Hmm, well it indeed does seem unlikely that (most) readers won't be able to make the connection, true, even though it is technically incorrect. However, even so, it still may appear to be a formatting mistake, it dilutes the well-established meaning of monospaced text in docs for readers, and results in a moment of cognitive dissonance for readers familiar with it that takes. These costs may be relatively trivial for a single instance, but readers constantly having to process such inconsistencies can add up to a non-trivial mental toll reducing overall docs quality, particularly for reference documentation such as this. Indeed, [Diataxis stresses this point repeatedly](https://diataxis.fr/reference/):

> Your users need reference material because they need truth and certainty - firm platforms on which to stand while they work.

> There should be no doubt or ambiguity in reference; it should be wholly authoritative.

> Reference material benefits from consistency. Be consistent, in structure, language, terminology, tone.

> These descriptions must be accurate [...] any discrepancy between the machinery and your description of it will inevitably lead a user astray.

Also, `dict` refers to the concrete Python _class_, while "dictionary" refers to the abstract Python data structure/protocol—which are separate references in the docs. If we want to refer to the latter instead here (and as prose), we can link [the appropriate glossary term](https://docs.python.org/3.11/glossary.html#term-dictionary) (and presumably should do the same for [list](https://docs.python.org/3.11/glossary.html#term-list):

```suggestion
:term:`dictionaries <dictionary>`),
```

Reviewer 2: In context `sequences (such tuple or list) or mappings (such as dict)`, dict is correct as an example parallel to list and tuple.  I am going to change 'like' to 'such as as better grammatically and parallel to the first parenthetical expression.

   Called when the instance is "called" as a function; if this method is defined,
   ``x(arg1, arg2, ...)`` roughly translates to ``type(x).__call__(x, arg1, ...)``.

>    The :class:`object` class itself does not provide this method.

Reviewer 4: Removed a blank line inadvertantly:

```suggestion
   The :class:`object` class itself does not provide this method.

```

.. _sequence-types:

Emulating container types
-------------------------

The following methods can be defined to implement container objects. None of them
are provided by the :class:`object` class itself. Containers usually are
:term:`sequences <sequence>` (such as :class:`lists <list>` or
:class:`tuples <tuple>`) or :term:`mappings <mapping>` (like

> :term:`dictionaries <dictionary>`),

Reviewer 4: 🤔 Hm…  on the line before, we see a general term used (with a link to glossary) and a concrete example in parens (list and tuple links), then we have the general/glossary term mappings and should keep the specific type dict in parens to keep the intent of the text.  On the other hand, if the glossary term for dictionaries is a better explanation than the dict type/function doc, and more specific than the term for mappings, the changed link could be an improvement.
